habitat/tasks/nav/greedy_follower.py

```diff
@@ -0,0 +1,210 @@
+
+        self._mode = (
+            "spath"
+            if getattr(sim, "straight_spath_points", None) is not None
+            else "grad"
+        )
+        self._return_one_hot = return_one_hot
+
+    def _get_return_value(
+        self, action: SimulatorActions
+    ) -> Union[SimulatorActions, np.array]:
+        if self._return_one_hot:
+            return action_to_one_hot(SIM_NAME_TO_ACTION[action.value])
+        else:
+            return action
+
+    def get_next_action(
+        self, goal_pos: np.array
+    ) -> Union[SimulatorActions, np.array]:
+        if (
+            np.linalg.norm(goal_pos - self._sim.get_agent_state().position)
+            <= self._goal_radius
+        ):
+            return self._get_return_value(SimulatorActions.STOP)
+
+        max_grad_dir = self._est_max_grad_dir(goal_pos)
+        if max_grad_dir is None:
+            return self._get_return_value(SimulatorActions.FORWARD)
+        return self.step_along_grad(max_grad_dir, goal_pos)
+
+    def step_along_grad(
+        self, grad_dir: np.quaternion, goal_pos: np.array
+    ) -> Union[SimulatorActions, np.array]:
+        current_state = self._sim.get_agent_state()
+        alpha = self._angle_between_quats(
+            grad_dir, v_to_q(current_state.rotation)
+        )
+        if alpha <= np.deg2rad(self._sim.config.TURN_ANGLE) + 1e-3:
+            return self._get_return_value(SimulatorActions.FORWARD)
+        else:
+            sim_action = SIM_NAME_TO_ACTION[SimulatorActions.LEFT.value]
+            self._sim.step(sim_action)
+            best_turn = (
+                SimulatorActions.LEFT
+                if (
+                    self._angle_between_quats(
+                        grad_dir, v_to_q(self._sim.get_agent_state().rotation)
+                    )
+                    < alpha
+                )
+                else SimulatorActions.RIGHT
+            )
+            self._reset_agent_state(current_state)
+
+            # Check if forward reduces geodesic distance
+            curr_dist = self._geo_dist(goal_pos)
+            self._sim.step(SIM_NAME_TO_ACTION[SimulatorActions.FORWARD.value])
+            new_dist = self._geo_dist(goal_pos)
+            new_pos = self._sim.get_agent_state().position
+            movement_size = np.sqrt(
+                np.sum(np.square(new_pos - current_state.position))
+            )
+            self._reset_agent_state(current_state)
+            if new_dist < curr_dist and movement_size / self._step_size > 0.99:
+                return self._get_return_value(SimulatorActions.FORWARD)
+            else:
+                return self._get_return_value(best_turn)
+
+    def _reset_agent_state(self, state: habitat_sim.AgentState) -> None:
+        self._sim.set_agent_state(
+            state.position, state.rotation, reset_sensors=False
+        )
+
+    @staticmethod
+    def _angle_between_quats(q1: np.quaternion, q2: np.quaternion) -> float:
```

Comment:
if this functions `_angle_between_quats ` and `_quat_from_two_vectors ` is general enough, it can be used to utils where we have angle utils.

Reply:
Done